### ISSUE TITLE
IsWindowAppearing in menu is true for 2 frames

### ISSUE THREAD

**Issue by codecat** [ocornut/imgui#3982]
**Version/Branch of Dear ImGui:**

Version: 1.83 WIP (commit `085cff2f`)
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_glfw.cpp (using [hello_imgui](https://github.com/pthom/hello_imgui), unsure how many modifications there are to vanilla imgui)
Operating System: Windows & Linux

**My Issue/Question:**

When I call `IsWindowAppearing` inside of a `BeginMenu`, it will be true for 2 frames instead of 1.

**Standalone, minimal, complete and verifiable example:**
```
if (ImGui::BeginMainMenuBar()) {
  if (ImGui::BeginMenu("Foo")) {
    if (ImGui::IsWindowAppearing()) {
      // This will print for 2 frames
      printf("Window is appearing\n");
    }
  }
  ImGui::EndMainMenuBar();
}
```

**Comment by ocornut** [comment]
I believe the reasoning what initially that on their first frame
- regular window would have zero or very small size
- tooltips and popups may inherit a recycled size
And performing manual layout in those condition and following a `IsWindowAppearing()` test may exhibit more issues. So somehow I always knew this would be the case that it returns true twice in a row on some windows (it is intentional), and I let that slip, and right now I don't have enough data to see if we could backtrack on this but it would be worth investigating.

First of all, I am curious if this is a problem for you (e.g. very slow computation triggered by this flag) or if you just noticed this and wanted to report it?

**Comment by codecat** [comment]
Both. I noticed it early on already and figured it probably wasn't a problem, but the menu population I was doing became a bit more computationally expensive. The menu in the case where I noticed this is still pretty fast right now, but I've had some stutters when opening that menu before I optimized it. Doing that computation for 2 frames is a bit wasteful I think.

**Comment by ocornut** [comment]
It is indeed wasteful, but seeing I don't have an immediate "right" solution just right now it's good to get a feel of how blocking it may be.

**Comment by codecat** [comment]
For what it's worth, currently for me it's not a blocker - the menus open fast enough.

**Comment by ocornut** [comment]
After further investigation I've decided to rework this code with 84e6fe4fc741983824955aaba6fb8203be6c778e effectively fixing this.

I've looking at other uses of `Appearing` (basically in `UpdateMouseMovingWindowEndFrame()` and `SetItemDefaultFocus()`) as well. I think it seems safe to make this change.

As far as I remember `IsWindowAppearing()` was exposed while working on a project that extensively used old columns with custom code manipulating column widths. That old columns code contained issue that made columns widths lossy on very small windows so there might have been an incentive for waiting a second frame until setting up some things. But right now I can't track more useful evidence that we need `IsWindowAppearing()` to return true twice in a row. If I am wrong I suppose we'll eventually find out.

(Ref to old commits related using `Appearing` and `Active/WasActive` from EndFrame / UpdateMouseMovingWindowEndFrame: 87ae40843c0b5c526edd9cdb62b072cd0bcc49ff and 235cca4f971cda95d2b14c2d599e98aac6a0ca5e)

**Comment by codecat** [comment]
Thanks so much for the quick fix! :smiling_face_with_three_hearts:

**Comment by ocornut** [comment]
Please note that fix pushed two days ago had a bug, now fixed by f0fc8228373ff627de3f453a6a5344e69d0b1dea
